File: backend/internal/handlers/reports.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/personal-finance-management/backend/internal/middleware"
	"github.com/personal-finance-management/backend/internal/services"
)

// ReportsHandler handles report-related HTTP requests
type ReportsHandler struct {
	dbService *services.DatabaseService
}

// NewReportsHandler creates a new reports handler
func NewReportsHandler(dbService *services.DatabaseService) *ReportsHandler {
	return &ReportsHandler{
		dbService: dbService,
	}
}

// GetMonthlySummary handles GET /api/reports/monthly-summary
func (h *ReportsHandler) GetMonthlySummary(c *gin.Context) {
	userID := middleware.MustGetUserID(c)

	// Parse query parameters
	monthStr := c.DefaultQuery("month", strconv.Itoa(int(time.Now().Month())))
	yearStr := c.DefaultQuery("year", strconv.Itoa(time.Now().Year()))

	month, err := strconv.Atoi(monthStr)
	if err != nil || month < 1 || month > 12 {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Invalid month parameter. Must be between 1 and 12",
		})
		return
	}

	year, err := strconv.Atoi(yearStr)
	if err != nil || year < 1900 || year > 2100 {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Invalid year parameter. Must be between 1900 and 2100",
		})
		return
	}

	// Get the monthly summary
	summary, err := h.dbService.Repositories.GetMonthlySummary(c.Request.Context(), userID, month, year)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to generate monthly summary",
			"details": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, summary)
}

// GetSpendingTrends handles GET /api/reports/spending-trends
func (h *ReportsHandler) GetSpendingTrends(c *gin.Context) {
	userID := middleware.MustGetUserID(c)

	// Parse query parameters
	monthsStr := c.DefaultQuery("months", "12")
	categoryID := c.Query("category_id")

	months, err := strconv.Atoi(monthsStr)
	if err != nil || months < 1 || months > 60 {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Invalid months parameter. Must be between 1 and 60",
		})
		return
	}

	var categoryPtr *string
	if categoryID != "" {
		categoryPtr = &categoryID
	}

	// Get spending trends
	trends, err := h.dbService.Repositories.GetSpendingTrends(c.Request.Context(), userID, categoryPtr, months)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to generate spending trends",
			"details": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, trends)
}

// GetCashFlow handles GET /api/reports/cash-flow
func (h *ReportsHandler) GetCashFlow(c *gin.Context) {
	userID := middleware.MustGetUserID(c)

	// Parse query parameters
	startDateStr := c.Query("start_date")
	endDateStr := c.Query("end_date")

	// Default to current month if no dates provided
	now := time.Now()
	defaultStart := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
	defaultEnd := defaultStart.AddDate(0, 1, 0).Add(-time.Second)

	var startDate, endDate time.Time
	var err error

	if startDateStr != "" {
		startDate, err = time.Parse("2006-01-02", startDateStr)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{
				"error": "Invalid start_date format. Use YYYY-MM-DD",
			})
			return
		}
	} else {
		startDate = defaultStart
	}

	if endDateStr != "" {
		endDate, err = time.Parse("2006-01-02", endDateStr)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{
				"error": "Invalid end_date format. Use YYYY-MM-DD",
			})
			return
		}
	} else {
		endDate = defaultEnd
	}

	// Validate date range
	if endDate.Before(startDate) {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "end_date must be after start_date",
		})
		return
	}

	// Limit the date range to prevent excessive queries
	if endDate.Sub(startDate) > 365*24*time.Hour {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Date range cannot exceed 365 days",
		})
		return
	}

	// Get cash flow data
	cashFlow, err := h.dbService.Repositories.GetCashFlow(c.Request.Context(), userID, startDate, endDate)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to generate cash flow report",
			"details": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, cashFlow)
}

// GetReportSummary handles GET /api/reports/summary
func (h *ReportsHandler) GetReportSummary(c *gin.Context) {
	userID := middleware.MustGetUserID(c)

	// Get current month summary
	now := time.Now()
	currentMonth := int(now.Month())
	currentYear := now.Year()

	monthlySummary, err := h.dbService.Repositories.GetMonthlySummary(c.Request.Context(), userID, currentMonth, currentYear)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to generate report summary",
			"details": err.Error(),
		})
		return
	}

	// Get spending trends for last 6 months
	trends, err := h.dbService.Repositories.GetSpendingTrends(c.Request.Context(), userID, nil, 6)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to get spending trends",
			"details": err.Error(),
		})
		return
	}

	// Get cash flow for current month
	startOfMonth := time.Date(currentYear, time.Month(currentMonth), 1, 0, 0, 0, 0, time.UTC)
	endOfMonth := startOfMonth.AddDate(0, 1, 0).Add(-time.Second)

	cashFlow, err := h.dbService.Repositories.GetCashFlow(c.Request.Context(), userID, startOfMonth, endOfMonth)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to get cash flow",
			"details": err.Error(),
		})
		return
	}

	// Create summary response
	summary := gin.H{
		"user_id":         userID,
		"generated_at":    time.Now(),
		"current_month":   monthlySummary,
		"spending_trends": trends,
		"cash_flow":       cashFlow,
	}

	c.JSON(http.StatusOK, summary)
}

// GetBudgetPerformance handles GET /api/reports/budget-performance
func (h *ReportsHandler) GetBudgetPerformance(c *gin.Context) {
	userID := middleware.MustGetUserID(c)

	// Parse query parameters
	monthStr := c.DefaultQuery("month", strconv.Itoa(int(time.Now().Month())))
	yearStr := c.DefaultQuery("year", strconv.Itoa(time.Now().Year()))

	month, err := strconv.Atoi(monthStr)
	if err != nil || month < 1 || month > 12 {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Invalid month parameter. Must be between 1 and 12",
		})
		return
	}

	year, err := strconv.Atoi(yearStr)
	if err != nil || year < 1900 || year > 2100 {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Invalid year parameter. Must be between 1900 and 2100",
		})
		return
	}

	// Calculate date range for the month
	startDate := time.Date(year, time.Month(month), 1, 0, 0, 0, 0, time.UTC)
	endDate := startDate.AddDate(0, 1, 0).Add(-time.Second)

	// Get all active budgets
	budgets, err := h.dbService.Repositories.GetBudgetsByUserID(c.Request.Context(), userID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Failed to get budgets",
			"details": err.Error(),
		})
		return
	}

	// Calculate performance for each budget
	var budgetPerformance []gin.H
	totalBudgeted := 0.0
	totalSpent := 0.0
	overBudgetCount := 0

	for _, budget := range budgets {
		// Get transactions for this budget's category in the date range
		transactions, err := h.dbService.Repositories.GetTransactionsByDateRange(c.Request.Context(), userID, startDate, endDate)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"error":   "Failed to get transactions",
				"details": err.Error(),
			})
			return
		}

		// Calculate spending for this category
		categorySpending := 0.0
		for _, transaction := range transactions {
			if transaction.CategoryID != nil && *transaction.CategoryID == budget.CategoryID && transaction.TransactionType == "expense" {
				categorySpending += transaction.Amount
			}
		}

		// Calculate performance metrics
		budgetAmount := budget.Amount
		remaining := budgetAmount - categorySpending
		percentageUsed := (categorySpending / budgetAmount) * 100
		isOverBudget := categorySpending > budgetAmount

		if isOverBudget {
			overBudgetCount++
		}

		totalBudgeted += budgetAmount
		totalSpent += categorySpending

		// Determine status
		status := "under_budget"
		if isOverBudget {
			status = "over_budget"
		} else if percentageUsed >= 80 {
			status = "at_risk"
		}

		budgetPerformance = append(budgetPerformance, gin.H{
			"budget_id":       budget.ID,
			"budget_name":     budget.Name,
			"category_id":     budget.CategoryID,
			"budgeted_amount": budgetAmount,
			"spent_amount":    categorySpending,
			"remaining":       remaining,
			"percentage_used": percentageUsed,
			"status":          status,
			"is_over_budget":  isOverBudget,
			"period":          budget.Period,
		})
	}

	// Calculate overall performance
	overallPerformance := gin.H{
		"total_budgeted":     totalBudgeted,
		"total_spent":        totalSpent,
		"total_remaining":    totalBudgeted - totalSpent,
		"overall_percentage": (totalSpent / totalBudgeted) * 100,
		"budgets_count":      len(budgets),
		"over_budget_count":  overBudgetCount,
		"on_track_count":     len(budgets) - overBudgetCount,
	}

	response := gin.H{
		"user_id":             userID,
		"month":               month,
		"year":                year,
		"budget_performance":  budgetPerformance,
		"overall_performance": overallPerformance,
		"generated_at":        time.Now(),
	}

	c.JSON(http.StatusOK, response)
}

File: backend-mern/routes/reports.js
import express from 'express';
import Transaction from '../models/Transaction.js';
import Budget from '../models/Budget.js';
import Category from '../models/Category.js';
import { protect } from '../middleware/auth.js';

const router = express.Router();

router.use(protect);

// @route   GET /api/reports/monthly-summary
// @desc    Get monthly income/expense summary
// @access  Private
router.get('/monthly-summary', async (req, res) => {
  try {
    const { month, year } = req.query;

    // Default to current month/year
    const now = new Date();
    const targetMonth = month ? parseInt(month) : now.getMonth() + 1;
    const targetYear = year ? parseInt(year) : now.getFullYear();

    // Validate parameters
    if (targetMonth < 1 || targetMonth > 12) {
      return res.status(400).json({ error: 'Invalid month parameter. Must be between 1 and 12' });
    }

    if (targetYear < 1900 || targetYear > 2100) {
      return res.status(400).json({ error: 'Invalid year parameter. Must be between 1900 and 2100' });
    }

    // Calculate date range for the month
    const startDate = new Date(targetYear, targetMonth - 1, 1);
    const endDate = new Date(targetYear, targetMonth, 0, 23, 59, 59);

    // Get all transactions for the month
    const transactions = await Transaction.find({
      userId: req.user._id,
      transactionDate: {
        $gte: startDate,
        $lte: endDate
      }
    }).populate('categoryId');

    // Calculate summary
    let totalIncome = 0;
    let totalExpense = 0;
    const categoryBreakdown = {};

    transactions.forEach(transaction => {
      const amount = parseFloat(transaction.amount);

      if (transaction.transactionType === 'income') {
        totalIncome += amount;
      } else if (transaction.transactionType === 'expense') {
        totalExpense += amount;
      }

      // Category breakdown
      if (transaction.categoryId) {
        const categoryName = transaction.categoryId.name;
        if (!categoryBreakdown[categoryName]) {
          categoryBreakdown[categoryName] = {
            income: 0,
            expense: 0,
            count: 0
          };
        }

        if (transaction.transactionType === 'income') {
          categoryBreakdown[categoryName].income += amount;
        } else {
          categoryBreakdown[categoryName].expense += amount;
        }
        categoryBreakdown[categoryName].count++;
      }
    });

    const netCashFlow = totalIncome - totalExpense;
    const savingsRate = totalIncome > 0 ? ((netCashFlow / totalIncome) * 100) : 0;

    res.json({
      month: targetMonth,
      year: targetYear,
      totalIncome: Math.round(totalIncome * 100) / 100,
      totalExpense: Math.round(totalExpense * 100) / 100,
      netCashFlow: Math.round(netCashFlow * 100) / 100,
      savingsRate: Math.round(savingsRate * 100) / 100,
      transactionCount: transactions.length,
      categoryBreakdown,
      generatedAt: new Date()
    });
  } catch (error) {
    console.error('Monthly summary error:', error);
    res.status(500).json({ error: 'Error generating monthly summary' });
  }
});

// @route   GET /api/reports/spending-trends
// @desc    Get spending trends over time
// @access  Private
router.get('/spending-trends', async (req, res) => {
  try {
    const { months = 12, category_id } = req.query;

    const monthsNum = parseInt(months);
    if (monthsNum < 1 || monthsNum > 60) {
      return res.status(400).json({ error: 'Invalid months parameter. Must be between 1 and 60' });
    }

    // Calculate date range
    const endDate = new Date();
    const startDate = new Date();
    startDate.setMonth(startDate.getMonth() - monthsNum);

    // Build query
    const query = {
      userId: req.user._id,
      transactionDate: {
        $gte: startDate,
        $lte: endDate
      }
    };

    if (category_id) {
      query.categoryId = category_id;
    }

    // Get transactions
    const transactions = await Transaction.find(query).populate('categoryId');

    // Group by month
    const monthlyData = {};

    transactions.forEach(transaction => {
      const date = new Date(transaction.transactionDate);
      const monthKey = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;

      if (!monthlyData[monthKey]) {
        monthlyData[monthKey] = {
          month: monthKey,
          income: 0,
          expense: 0,
          net: 0,
          transactionCount: 0
        };
      }

      const amount = parseFloat(transaction.amount);

      if (transaction.transactionType === 'income') {
        monthlyData[monthKey].income += amount;
      } else if (transaction.transactionType === 'expense') {
        monthlyData[monthKey].expense += amount;
      }

      monthlyData[monthKey].net = monthlyData[monthKey].income - monthlyData[monthKey].expense;
      monthlyData[monthKey].transactionCount++;
    });

    // Convert to array and sort
    const trends = Object.values(monthlyData)
      .sort((a, b) => a.month.localeCompare(b.month))
      .map(data => ({
        ...data,
        income: Math.round(data.income * 100) / 100,
        expense: Math.round(data.expense * 100) / 100,
        net: Math.round(data.net * 100) / 100
      }));

    // Calculate averages
    const avgIncome = trends.length > 0
      ? trends.reduce((sum, t) => sum + t.income, 0) / trends.length
      : 0;
    const avgExpense = trends.length > 0
      ? trends.reduce((sum, t) => sum + t.expense, 0) / trends.length
      : 0;

    res.json({
      trends,
      summary: {
        averageIncome: Math.round(avgIncome * 100) / 100,
        averageExpense: Math.round(avgExpense * 100) / 100,
        averageNet: Math.round((avgIncome - avgExpense) * 100) / 100,
        monthsAnalyzed: trends.length
      },
      generatedAt: new Date()
    });
  } catch (error) {
    console.error('Spending trends error:', error);
    res.status(500).json({ error: 'Error generating spending trends' });
  }
});

// @route   GET /api/reports/cash-flow
// @desc    Get cash flow report for date range
// @access  Private
router.get('/cash-flow', async (req, res) => {
  try {
    const { start_date, end_date } = req.query;

    // Default to current month if no dates provided
    const now = new Date();
    const defaultStart = new Date(now.getFullYear(), now.getMonth(), 1);
    const defaultEnd = new Date(now.getFullYear(), now.getMonth() + 1, 0, 23, 59, 59);

    let startDate, endDate;

    if (start_date) {
      startDate = new Date(start_date);
      if (isNaN(startDate.getTime())) {
        return res.status(400).json({ error: 'Invalid start_date format. Use YYYY-MM-DD' });
      }
    } else {
      startDate = defaultStart;
    }

    if (end_date) {
      endDate = new Date(end_date);
      if (isNaN(endDate.getTime())) {
        return res.status(400).json({ error: 'Invalid end_date format. Use YYYY-MM-DD' });
      }
    } else {
      endDate = defaultEnd;
    }

    // Validate date range
    if (endDate < startDate) {
      return res.status(400).json({ error: 'end_date must be after start_date' });
    }

    const daysDiff = (endDate - startDate) / (1000 * 60 * 60 * 24);
    if (daysDiff > 365) {
      return res.status(400).json({ error: 'Date range cannot exceed 365 days' });
    }

    // Get transactions
    const transactions = await Transaction.find({
      userId: req.user._id,
      transactionDate: {
        $gte: startDate,
        $lte: endDate
      }
    }).populate('categoryId').sort({ transactionDate: 1 });

    // Calculate daily cash flow
    const dailyCashFlow = {};
    let runningBalance = 0;

    transactions.forEach(transaction => {
      const dateKey = transaction.transactionDate.toISOString().split('T')[0];

      if (!dailyCashFlow[dateKey]) {
        dailyCashFlow[dateKey] = {
          date: dateKey,
          income: 0,
          expense: 0,
          net: 0,
          balance: 0,
          transactions: []
        };
      }

      const amount = parseFloat(transaction.amount);

      if (transaction.transactionType === 'income') {
        dailyCashFlow[dateKey].income += amount;
        runningBalance += amount;
      } else if (transaction.transactionType === 'expense') {
        dailyCashFlow[dateKey].expense += amount;
        runningBalance -= amount;
      }

      dailyCashFlow[dateKey].net = dailyCashFlow[dateKey].income - dailyCashFlow[dateKey].expense;
      dailyCashFlow[dateKey].balance = runningBalance;
      dailyCashFlow[dateKey].transactions.push({
        id: transaction._id,
        description: transaction.description,
        amount: amount,
        type: transaction.transactionType,
        category: transaction.categoryId?.name || 'Uncategorized'
      });
    });

    // Convert to array and sort
    const cashFlowData = Object.values(dailyCashFlow)
      .sort((a, b) => a.date.localeCompare(b.date))
      .map(data => ({
        ...data,
        income: Math.round(data.income * 100) / 100,
        expense: Math.round(data.expense * 100) / 100,
        net: Math.round(data.net * 100) / 100,
        balance: Math.round(data.balance * 100) / 100
      }));

    // Calculate totals
    const totalIncome = cashFlowData.reduce((sum, d) => sum + d.income, 0);
    const totalExpense = cashFlowData.reduce((sum, d) => sum + d.expense, 0);

    res.json({
      startDate: startDate.toISOString().split('T')[0],
      endDate: endDate.toISOString().split('T')[0],
      dailyCashFlow: cashFlowData,
      summary: {
        totalIncome: Math.round(totalIncome * 100) / 100,
        totalExpense: Math.round(totalExpense * 100) / 100,
        netCashFlow: Math.round((totalIncome - totalExpense) * 100) / 100,
        finalBalance: cashFlowData.length > 0 ? cashFlowData[cashFlowData.length - 1].balance : 0,
        daysAnalyzed: Object.keys(dailyCashFlow).length
      },
      generatedAt: new Date()
    });
  } catch (error) {
    console.error('Cash flow error:', error);
    res.status(500).json({ error: 'Error generating cash flow report' });
  }
});

// @route   GET /api/reports/summary
// @desc    Get comprehensive report summary
// @access  Private
router.get('/summary', async (req, res) => {
  try {
    const now = new Date();
    const currentMonth = now.getMonth() + 1;
    const currentYear = now.getFullYear();

    // Calculate date ranges
    const startOfMonth = new Date(currentYear, currentMonth - 1, 1);
    const endOfMonth = new Date(currentYear, currentMonth, 0, 23, 59, 59);
    const sixMonthsAgo = new Date();
    sixMonthsAgo.setMonth(sixMonthsAgo.getMonth() - 6);

    // Get current month transactions
    const monthlyTransactions = await Transaction.find({
      userId: req.user._id,
      transactionDate: {
        $gte: startOfMonth,
        $lte: endOfMonth
      }
    }).populate('categoryId');

    // Calculate monthly summary
    let monthlyIncome = 0;
    let monthlyExpense = 0;

    monthlyTransactions.forEach(t => {
      const amount = parseFloat(t.amount);
      if (t.transactionType === 'income') monthlyIncome += amount;
      else if (t.transactionType === 'expense') monthlyExpense += amount;
    });

    // Get 6-month trends
    const trendTransactions = await Transaction.find({
      userId: req.user._id,
      transactionDate: {
        $gte: sixMonthsAgo,
        $lte: now
      }
    });

    const monthlyTrends = {};
    trendTransactions.forEach(t => {
      const date = new Date(t.transactionDate);
      const monthKey = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;

      if (!monthlyTrends[monthKey]) {
        monthlyTrends[monthKey] = { income: 0, expense: 0 };
      }

      const amount = parseFloat(t.amount);
      if (t.transactionType === 'income') monthlyTrends[monthKey].income += amount;
      else if (t.transactionType === 'expense') monthlyTrends[monthKey].expense += amount;
    });

    res.json({
      userId: req.user._id.toString(),
      generatedAt: now,
      currentMonth: {
        month: currentMonth,
        year: currentYear,
        totalIncome: Math.round(monthlyIncome * 100) / 100,
        totalExpense: Math.round(monthlyExpense * 100) / 100,
        netCashFlow: Math.round((monthlyIncome - monthlyExpense) * 100) / 100,
        transactionCount: monthlyTransactions.length
      },
      spendingTrends: Object.entries(monthlyTrends)
        .map(([month, data]) => ({
          month,
          income: Math.round(data.income * 100) / 100,
          expense: Math.round(data.expense * 100) / 100,
          net: Math.round((data.income - data.expense) * 100) / 100
        }))
        .sort((a, b) => a.month.localeCompare(b.month)),
      cashFlow: {
        totalIncome: Math.round(monthlyIncome * 100) / 100,
        totalExpense: Math.round(monthlyExpense * 100) / 100,
        netCashFlow: Math.round((monthlyIncome - monthlyExpense) * 100) / 100
      }
    });
  } catch (error) {
    console.error('Report summary error:', error);
    res.status(500).json({ error: 'Error generating report summary' });
  }
});

// @route   GET /api/reports/budget-performance
// @desc    Get budget vs actual spending performance
// @access  Private
router.get('/budget-performance', async (req, res) => {
  try {
    const { month, year } = req.query;

    const now = new Date();
    const targetMonth = month ? parseInt(month) : now.getMonth() + 1;
    const targetYear = year ? parseInt(year) : now.getFullYear();

    // Validate parameters
    if (targetMonth < 1 || targetMonth > 12) {
      return res.status(400).json({ error: 'Invalid month parameter. Must be between 1 and 12' });
    }

    if (targetYear < 1900 || targetYear > 2100) {
      return res.status(400).json({ error: 'Invalid year parameter. Must be between 1900 and 2100' });
    }

    // Calculate date range
    const startDate = new Date(targetYear, targetMonth - 1, 1);
    const endDate = new Date(targetYear, targetMonth, 0, 23, 59, 59);

    // Get all active budgets for user
    const budgets = await Budget.find({ userId: req.user._id }).populate('categoryId');

    // Get all expense transactions for the period
    const transactions = await Transaction.find({
      userId: req.user._id,
      transactionType: 'expense',
      transactionDate: {
        $gte: startDate,
        $lte: endDate
      }
    });

    // Calculate spending by category
    const categorySpending = {};
    transactions.forEach(t => {
      const categoryId = t.categoryId?.toString();
      if (categoryId) {
        categorySpending[categoryId] = (categorySpending[categoryId] || 0) + parseFloat(t.amount);
      }
    });

    // Analyze each budget
    const budgetPerformance = [];
    let totalBudgeted = 0;
    let totalSpent = 0;
    let overBudgetCount = 0;

    for (const budget of budgets) {
      const categoryId = budget.categoryId?._id?.toString();
      const budgetAmount = parseFloat(budget.amount);
      const spentAmount = categorySpending[categoryId] || 0;
      const remaining = budgetAmount - spentAmount;
      const percentageUsed = budgetAmount > 0 ? (spentAmount / budgetAmount) * 100 : 0;
      const isOverBudget = spentAmount > budgetAmount;

      if (isOverBudget) overBudgetCount++;

      totalBudgeted += budgetAmount;
      totalSpent += spentAmount;

      // Determine status
      let status = 'under_budget';
      if (isOverBudget) {
        status = 'over_budget';
      } else if (percentageUsed >= 80) {
        status = 'at_risk';
      }

      budgetPerformance.push({
        budgetId: budget._id,
        budgetName: budget.categoryId?.name || 'Unknown',
        categoryId: categoryId,
        budgetedAmount: Math.round(budgetAmount * 100) / 100,
        spentAmount: Math.round(spentAmount * 100) / 100,
        remaining: Math.round(remaining * 100) / 100,
        percentageUsed: Math.round(percentageUsed * 100) / 100,
        status,
        isOverBudget,
        period: budget.period || 'monthly'
      });
    }

    // Calculate overall performance
    const overallPerformance = {
      totalBudgeted: Math.round(totalBudgeted * 100) / 100,
      totalSpent: Math.round(totalSpent * 100) / 100,
      totalRemaining: Math.round((totalBudgeted - totalSpent) * 100) / 100,
      overallPercentage: totalBudgeted > 0 ? Math.round((totalSpent / totalBudgeted) * 100 * 100) / 100 : 0,
      budgetsCount: budgets.length,
      overBudgetCount,
      onTrackCount: budgets.length - overBudgetCount
    };

    res.json({
      userId: req.user._id.toString(),
      month: targetMonth,
      year: targetYear,
      budgetPerformance,
      overallPerformance,
      generatedAt: new Date()
    });
  } catch (error) {
    console.error('Budget performance error:', error);
    res.status(500).json({ error: 'Error generating budget performance report' });
  }
});

export default router;
